bee-vote/src/context.rs

```diff
@@ -0,0 +1,252 @@
+    pub fn message_id(&self) -> Option<MessageId> {
+        if let Self::Timestamp(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+
+    /// Get a `TransactionId` if this is a Conflict object.
+    /// Returns `Some(id)` if this is the canse, `None` if not.
+    pub fn transaction_id(&self) -> Option<TransactionId> {
+        if let Self::Conflict(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+}
+
+/// Builder pattern struct for instantiating a `VoteContext`.
+pub struct VoteContextBuilder {
+    /// Object of the voting (conflict or timestamp), and associated ID.
+    object: VoteObject,
+    /// Opinions held by this context on the vote object.
+    opinions: Option<Opinions>,
+}
+
+impl VoteContextBuilder {
+    /// Create a new `VoteContextBuilder`, defining an ID and an `ObjectType` (voting object).
+    pub fn new(object: VoteObject) -> Self {
+        Self { object, opinions: None }
+    }
+
+    /// Set a single initial `Opinion`.
+    pub fn with_initial_opinion(mut self, opinion: Opinion) -> Self {
+        self.opinions = Some(Opinions::new(vec![opinion]));
+        self
+    }
+
+    /// Set a list of initial `Opinion`s.
+    pub fn with_initial_opinions(mut self, opinions: Opinions) -> Self {
+        self.opinions = Some(opinions);
+        self
+    }
+
+    /// Build a `VoteContext`.
+    /// Note: this will panic if no initial opinions have been provided.
+    pub fn build(self) -> Result<VoteContext, Error> {
+        Ok(VoteContext {
+            object: self.object,
+            liked: LIKED_INITIAL,
+            rounds: 0,
+            opinions: self.opinions.ok_or(Error::NoInitialOpinions)?,
+        })
+    }
+}
+
+/// Voting context.
+#[derive(Debug, Clone)]
+pub struct VoteContext {
+    /// Object type of the vote and related object ID.
+    object: VoteObject,
+    /// The percentage of `OpinionGiver`s who liked this item on the last query.
+    liked: f64,
+    /// The number of voting rounds performed so far.
+    rounds: u32,
+    /// List of opinions formed at the end of each voting round.
+    /// The first in this list is the initial opinion when this `VoteContext` was created.
+    opinions: Opinions,
+}
+
+impl VoteContext {
+    /// Constructs a new `VoteContext`.
+    pub(crate) fn new(object: VoteObject, initial_opinion: Opinion) -> Self {
+        Self {
+            object,
+            liked: LIKED_INITIAL,
+            rounds: 0,
+            opinions: Opinions::new(vec![initial_opinion]),
+        }
+    }
+
+    /// Add the given `Opinion` to the `VoteContext`.
+    pub(crate) fn add_opinion(&mut self, opinion: Opinion) {
+        self.opinions.push(opinion);
+    }
+
+    /// Retrieve the last formed `Opinion`.
+    pub fn last_opinion(&self) -> Option<Opinion> {
+        self.opinions.last().copied()
+    }
+
+    /// Describes whether this `VoteContext` has been finalized.
```

Comment:
Can you also add a comment on how it is determined when a vote is finalized (same opinion for n rounds) and the implications (will not change it's opinion in the future).

Reply:
Yep, I'll add comments for each of the areas highlighted in your suggestions 👍